### PR TITLE
Redcore part of AESEC-3637

### PR DESCRIPTION
**Pull request by atabegruslan**

### CODE REVIEW

File: extensions/libraries/redcore/api/payment/plugin/helper/payment.php [hunk 1]
```diff
@@ -444,11 +444,13 @@ protected function getPaymentByExtensionOrderData($extensionName, $orderData)
 		$payment = RApiPaymentHelper::getPaymentByExtensionId($extensionName, $orderId);
 		$restrictedData = array(
 			'id', 'extension_name', 'payment_name', 'sandbox', 'created_date', 'modified_date', 'confirmed_date', 'transaction_id',
-			'amount_total', 'amount_paid', 'coupon_code', 'customer_note', 'status'
+			'amount_paid', 'coupon_code', 'customer_note', 'status'
```
Comment (atabegruslan) [inline extensions/libraries/redcore/api/payment/plugin/helper/payment.php]: The 3 alterations here to function `getPaymentByExtensionOrderData` is to fix `amount_total` & `amount_original` being the wrong-way-around after order's total cost is changed.

https://redweb.slack.com/archives/C02951NC0/p1534236714000100
-
https://redweb.slack.com/archives/C02951NC0/p1534238084000100

File: extensions/libraries/redcore/api/payment/helper.php [hunk 2]
```diff
@@ -759,6 +763,26 @@ public static function updatePaymentStatus($paymentId)
...
+	 * @param   object	$payment	Payment object
+	 *
+	 * @return  boolean
+	 */
+	public static function isInstantCapture($payment)
+	{
+		$plugin 		= RApiPaymentHelper::getPaymentParams($payment->payment_name, $payment->extension_name, $payment->owner_name);
+		$instantcapture = false;
+
+		if ($plugin && strcmp($plugin->element, 'epay') === 0)
```
Comment (Kixo) [inline extensions/libraries/redcore/api/payment/helper.php]: maybe not setting it strictly to epay plugin... other payment plugins can have partial payment as well...

Comment (atabegruslan) [reply extensions/libraries/redcore/api/payment/helper.php]: We have:
   paypal
   epay
   bank transfer
   ean
   quickpay 

Can I please know to which ones this should be applied to?

I just know that qPay and paypal have. I dont think bank transfer have but I am not sure. I dont know what ean and quickspay are.

File: extensions/libraries/redcore/api/payment/helper.php [hunk 3]
```diff
@@ -759,6 +763,27 @@ public static function updatePaymentStatus($paymentId)
...
+	 *
+	 * @return  boolean
+	 */
+	public static function isInstantCapture($payment)
+	{
+		$plugin 		= RApiPaymentHelper::getPaymentParams($payment->payment_name, $payment->extension_name, $payment->owner_name);
+		$instantcapture = false;
+		$onlinePayments = array('epay', 'paypal', 'quickpay');
+
+		if ($plugin && in_array($plugin->element, $onlinePayments))
```
Comment (Kixo) [inline extensions/libraries/redcore/api/payment/helper.php]: I am not sure that in_array($plugin->element, $onlinePayments) is needed if you have fixed $plugin->params->get('instantcapture', 0) parameter that we will use in all our plugins.

Comment (atabegruslan) [reply extensions/libraries/redcore/api/payment/helper.php]: @Kixo But Paypal and Quickpay also can do partial capture, but those 2 plugins don't have the `instantcapture` parameter.

Comment (atabegruslan) [reply extensions/libraries/redcore/api/payment/helper.php]: Actually I didn't understand your "...if you have fixed $plugin->params->get('instantcapture', 0) parameter that we will use in all our plugins."
Did you mean that the `if (in_array($plugin->element, $onlinePayments))` check around `$instantcapture = (boolean) $plugin->params->get('instantcapture', 0);` is unnecesary and that `$instantcapture = (boolean) $plugin->params->get('instantcapture', 0);` by itself is fine?

Comment (atabegruslan) [reply extensions/libraries/redcore/api/payment/helper.php]: I tested with bank transfer. In order, I still have the capture payment option :(

Comment (atabegruslan) [reply extensions/libraries/redcore/api/payment/helper.php]: image
Paypal can handle capture payments and partial captures, yet I get this error :(

I think I just discovered a new load of problems

**Comment by atabegruslan** [comment]
@Kixo Regarding https://github.com/redCOMPONENT-COM/redCORE/pull/841#discussion_r222688037
I changed it back to `if ($plugin && strcmp($plugin->element, 'epay') === 0)`.

It works for ePay. The other pay plugins isn't yet ready to handle "manual capturing".
I found out:
- Paypal have partial capture, but our paypal plugin cant even handle payment capture. Maybe later, someone can add implement `function handleCapturePayment` in `plugins/redpayment/paypal/helpers/payment.php`.
- Quickpay have partial capture, but we hardly use it. Redweb only used it once in a RedShop site.
- Bank transfer strangely implements `handleCapturePayment` https://gitlab.redhost.dk/redweb-extensions/redCOREPLUGINS/blob/develop/extensions/plugins/redpayment/bank_transfer/helpers/payment.php#L180 . Maybe later someone can delete that.
- Lastly, only our epay plugin have the instant_capture parameter, though I dont think epay is the only payment method that have instant capture.

The plugins are in a mess. So I am going to leave this in a state, where this will completely work with epay (the only payment plugin that is ready for partial capture)

File: extensions/libraries/redcore/api/payment/helper.php (hunk 2 above)
Comment (shandak) [reply extensions/libraries/redcore/api/payment/helper.php]: better not to use a specific plugin name but some kind of flag. For instance, you can try to check some object plugin property value or plugin param like `if ($params->get('hasSomeSpecificAction', false))` ?

Comment (atabegruslan) [reply extensions/libraries/redcore/api/payment/helper.php]: @shandak Understood. I was told about that earlier too. 
But thing are quite messy and currently only ePay can handle capture.
You can see my other comments here.

**Pull request merged**